Fix deceleration in CharacterBody 2D and 3D templates

ubitux opened the pull request:
https://github.com/godotengine/godot/pull/73873 is stalled for more than a year, so I thought addressing small but concrete parts of the problem could help making progress.

*Physics squad edit:*

- Fixes https://github.com/godotengine/godot/issues/74216

AThousandShips commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -19,9 +19,6 @@ func _physics_process(delta: float) -> void:
-		velocity.x = direction * SPEED
-	else:
-		velocity.x = move_toward(velocity.x, 0, SPEED)
+	velocity.x = direction * SPEED
```
Why this change? It ignores the deceleration, this significantly changes the behaviour, which isn't a "clean up"

ubitux replied:
It makes no difference to the behavior; can you check if the commit description helps you understand the issue?

AThousandShips replied:
No it absolutely changes it, I've used that code, how did you establish it doesn't?

AThousandShips replied:
>  move_toward calls are equivalent to
assigning the velocity components to 0

This is false, *unless* the correct fix is to do `SPEED * delta` for the method

ubitux replied:
`velocity` can never be greater than speed because `direction` is normalized, which means the code is equivalent to:
```
if direction:
    velocity.x = direction * SPEED
else:
    velocity.x = 0
```

Which itself is equivalent to `velocity.x = direction * SPEED`

AThousandShips replied:
Then the correct code is `move_toward(velocity.x, 0, SPEED * delta)`

ubitux replied:
I updated the description with a clearer wording

ubitux replied:
Then this implies a lot more changes because with the current `SPEED` the friction is way too low, and we should probably split that constant `SPEED` into a deceleration constant.

I thought this kind of improvement to the code could be added on top of this PR which is meant to clarify the current state of the code (admittedly loosing the original intention of the code, which I couldn't get looking at it)

AThousandShips replied:
> I thought this kind of improvement

It's not an improvement if the code is broken, and if it is then the bug should be fixed, if anything should be done, otherwise I don't think there's anything in this PR which significantly improves the state of this, and should be done in a PR fixing the code instead

AThousandShips replied:
Otherwise I'll just go ahead and fix this tomorrow, if you don't feel confident doing so

ubitux replied:
Commit replaced with the introduction of a deceleration constant

AThousandShips commented on modules/mono/editor/script_templates/CharacterBody2D/basic_movement.cs:
```diff
@@ -27,14 +27,7 @@ public override void _PhysicsProcess(double delta)
-        {
-            velocity.X = Mathf.MoveToward(Velocity.X, 0, Speed);
-        }
+        velocity.X = direction.X * Speed;
```
Same here

AThousandShips commented on modules/mono/editor/script_templates/CharacterBody3D/basic_movement.cs:
```diff
@@ -28,16 +28,8 @@ public override void _PhysicsProcess(double delta)
-            velocity.X = Mathf.MoveToward(Velocity.X, 0, Speed);
-            velocity.Z = Mathf.MoveToward(Velocity.Z, 0, Speed);
-        }
+        velocity.X = direction.X * Speed;
```
And here

AThousandShips commented on modules/gdscript/editor/script_templates/CharacterBody3D/basic_movement.gd:
```diff
@@ -20,11 +20,7 @@ func _physics_process(delta: float) -> void:
-		velocity.x = move_toward(velocity.x, 0, SPEED)
-		velocity.z = move_toward(velocity.z, 0, SPEED)
+	velocity.x = direction.x * SPEED
+	velocity.z = direction.z * SPEED
```
And here

ubitux committed:
Fix deceleration in CharacterBody 2D and 3D templates

The velocity can never be greater than SPEED because direction is
normalized. This means the move_toward calls are currently equivalent to
assigning the velocity components to 0.

Changing SPEED to SPEED * delta makes the friction effect way too small
(the character drifts too much) so we introduce a new deceleration
constant.

The cosmetics regarding zeros written as floating point are unrelated
to the fix itself.

rburing commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -3,8 +3,9 @@
 
-const SPEED = 300.0
-const JUMP_VELOCITY = -400.0
+const SPEED : float = 300.0
```
I think we should follow the [Writing guidelines](https://docs.godotengine.org/en/stable/contributing/documentation/docs_writing_guidelines.html#dynamic-vs-static-typing) and omit the types. (Even though I am a static types enjoyer.)

A nice improvement to the template system (separate from this PR) would be to offer a variant with type hints depending on the user's settings.

ubitux replied:
I was thinking that with round numbers like this, it makes more sense to secure it with the explicit float rather than depending on `:=` in case this ends up being changed to a round integer constant `300` which may cause weird bugs down the line

I can also drop the commit and keep the current `=`

dalexeev replied:
This will be preprocessed here:

https://github.com/godotengine/godot/blob/be56cab58c056c074d1e02cd0b38641204e39f41/modules/gdscript/gdscript_editor.cpp#L78-L96

ubitux replied:
Commit removed

rburing commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -3,8 +3,9 @@
-const SPEED = 300.0
-const JUMP_VELOCITY = -400.0
+const SPEED : float = 300.0
+const DECEL : float = 1000.0
```
I think `DECEL` can be confusing to non-native English speakers, so it should be e.g. `DECELERATION`.

ubitux replied:
Renamed

dalexeev commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -3,8 +3,9 @@
 
-const SPEED = 300.0
-const JUMP_VELOCITY = -400.0
+const SPEED : float = 300.0
```
```suggestion
const SPEED: float = 300.0
```

The same below.

ubitux replied:
Fixed

raulsntos commented on modules/mono/editor/script_templates/CharacterBody2D/basic_movement.cs:
```diff
@@ -33,7 +34,7 @@ public override void _PhysicsProcess(double delta)
         else
         {
-            velocity.X = Mathf.MoveToward(Velocity.X, 0, Speed);
+            velocity.X = Mathf.MoveToward(Velocity.X, 0, Deceleration * delta);
```
```suggestion
            velocity.X = Mathf.MoveToward(Velocity.X, 0, Deceleration * (float)delta);
```

ubitux replied:
done

raulsntos commented on modules/mono/editor/script_templates/CharacterBody3D/basic_movement.cs:
```diff
@@ -35,8 +36,8 @@ public override void _PhysicsProcess(double delta)
         {
-            velocity.X = Mathf.MoveToward(Velocity.X, 0, Speed);
-            velocity.Z = Mathf.MoveToward(Velocity.Z, 0, Speed);
+            velocity.X = Mathf.MoveToward(Velocity.X, 0, Deceleration * delta);
+            velocity.Z = Mathf.MoveToward(Velocity.Z, 0, Deceleration * delta);
```
```suggestion
            velocity.X = Mathf.MoveToward(Velocity.X, 0, Deceleration * (float)delta);
            velocity.Z = Mathf.MoveToward(Velocity.Z, 0, Deceleration * (float)delta);
```

ubitux replied:
done

dalexeev reviewed (commented):
The GDScript code style looks good to me, but I think it's up to the physics team to decide.

rburing commented on modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd:
```diff
@@ -22,6 +23,6 @@ func _physics_process(delta: float) -> void:
 		velocity.x = direction * SPEED
 	else:
-		velocity.x = move_toward(velocity.x, 0, SPEED)
+		velocity.x = move_toward(velocity.x, 0, DECELERATION * delta)
```
Let's replace the literal `0` by `0.0` everywhere.

ubitux replied:
Done in a separate commit as it is unrelated to this change. I also replaced the other `0` for the input vector swizzling, I hope that's ok?

AThousandShips replied:
No need to keep separate commits, you'll have to squash into one at the end anyway

ubitux replied:
Just curious, is there an issue with keeping the 2 independent and functional commits?

AThousandShips replied:
Because we keep things to one commit when ever possible, to not clutter things and makes managing things easier, what's the issue with having just one commit for one change?

ubitux replied:
> what's the issue with having just one commit for one change?

It's not a single change since it's cosmetics on top of a fix, and I'm touching code that's unrelated to the fix itself (for example the input vector).

Also, It makes reverting or cherry-picking only the fix harder. It is also confusing to see mixed changes along with a fix, as it makes you question if the `0` vs `0.0` was actually part of the problem

AThousandShips replied:
For these purposes it's a single change, both are part of the same change, if the two need to be separate they should be two PRs instead, the changes aren't going to be reverted separately as the change to float constants is perfectly safe

ubitux replied:
Your call, squashed 🤷

ubitux commented:
> The bugfix this PR does is definitely welcome. I'm not completely sure about the name "deceleration" though, since it is a number and not a vector. (But maybe simplicity is more important than mathematical accuracy.)
> 
> Sorry for all the nitpicking but this code will be seen by many first-time users, so we want to get it right.

I don't mind renaming that, what do you suggest? Slowdown? MomentumForce?

Changing into a vector feels like it's going to open a can of worms to me, but your call.

AThousandShips commented:
> Changing into a vector feels like it's going to open a can of worms to me, but your call.

I think the point is that deceleration isn't a scalar but a vector, so the name not the type is in question

ubitux commented:
> > Changing into a vector feels like it's going to open a can of worms to me, but your call.
> 
> I think the point is that deceleration isn't a scalar but a vector, so the name not the type is in question

Sure, I'm happy to rename to anything, please let me know what you'd prefer.

ubitux commented:
Also, alternatively, we could use the incoming `move_toward_smooth()` (https://github.com/godotengine/godot/pull/92236) unconditionally; something like:
```gdscript
	velocity.x = move_toward_smooth(velocity.x, direction.x * SPEED, delta * 10.0)
	velocity.z = move_toward_smooth(velocity.z, direction.z * SPEED, delta * 10.0)
```
This handles both the acceleration and the deceleration at the same time and actually simplifies the code by removing the branching

AThousandShips commented:
That'd be something for a separate PR, would be good to get this fix in 4.3 and the other method won't be in 4.3

patwork commented:
https://github.com/godotengine/godot/blob/fc6d8e6338aa9d207a73bea5bb452dd8a9c6eff3/modules/gdscript/editor/script_templates/CharacterBody2D/basic_movement.gd#L22-L26

Why do we even need smooth deceleration here if there is no smooth acceleration? Changing velocity.x from 0 to SPEED is instantaneous.

Why not just https://docs.godotengine.org/en/latest/tutorials/physics/using_character_body_2d.html#platformer-movement 

```
	# Get the input direction.
	var direction = Input.get_axis("ui_left", "ui_right")
	velocity.x = direction * speed
```

patwork commented:
> if anything I'd say to add acceleration as well

Even given that this is intended to be a simple character controller, making available the possibility to set both acceleration and deceleration rates seems preferable.

Example from kidscancode, noteworthy also for showing how to enable modification of values in the inspector (@export).

https://kidscancode.org/godot_recipes/4.x/2d/platform_character/index.html#friction-and-acceleration

ubitux commented:
The point of this PR is to fix the code that has been broken for years, not rewrite it. A PR to rewrite it has been stalled forever, so the idea here is that making the code correct in the first place so that it can later be improved starting from a correct foundation.

There are many opinions on how the default template could be, and this could take a very long time before reaching a consensus, but I don't think it should prevent important fixes in the meantime.
